docarray/array/array/pushpull/__init__.py

```diff
@@ -0,0 +1,244 @@
+            logging.info('Loaded Local Filesystem backend')
+        elif protocol == 's3':
+            from docarray.array.array.pushpull.s3 import PushPullS3
+
+            cls.__backends__[protocol] = PushPullS3
+            logging.info('Loaded S3 backend')
+        else:
+            raise NotImplementedError(f'protocol {protocol} not supported')
+
+        return cls.__backends__[protocol]
+
+    @staticmethod
+    def list(
+        url: str = f'file://{__cache_path__}', show_table: bool = False
+    ) -> List[str]:
+        """
+        List all the DocumentArrays in the namespace.
+        url should be of the form ``protocol://namespace``
+
+        If no url is provided, the DocumentArrays in the local cache will be listed.
+
+        :param url: should be of the form ``protocol://namespace``. e.g. ``s3://bucket/path/to/namespace``, ``file:///path/to/folder``
+        :param show_table: whether to show the table of artifacts
+        :return: a list of artifact names
+        """
+        logging.info(f'Listing artifacts from {url}')
+        protocol, namespace = url.split('://', 2)
+        return PushPullMixin.get_backend(protocol).list(namespace, show_table)
+
+    @classmethod
+    def delete(cls, url: str, missing_ok: bool = False):
+        """
+        Delete the DocumentArray at the given url.
+
+        :param url: should be of the form ``protocol://namespace/name``. e.g. ``s3://bucket/path/to/namespace/name``, ``file:///path/to/folder/name``
+        :param missing_ok: whether to ignore if the artifact does not exist
+        """
+        logging.info(f'Deleting artifact {url}')
+        protocol, name = url.split('://', 2)
+        success = PushPullMixin.get_backend(protocol).delete(
+            name, missing_ok=missing_ok
+        )
+        if success:
+            logging.info(f'Successfully deleted artifact {url}')
+        else:
+            logging.warning(f'Failed to delete artifact {url}')
+        return success
+
+    def push(
+        self,
+        url: str,
+        public: bool = True,
+        show_progress: bool = False,
+        branding: Optional[Dict] = None,
+    ) -> Dict:
+        """Push this DocumentArray object to the specified url.
+
+        :param url: url specifying the protocol and save name of the DocumentArray. Should be of the form ``protocol://namespace/name``. e.g. ``s3://bucket/path/to/namespace/name``, ``file:///path/to/folder/name``
+        :param public:  Only used by ``jinaai`` protocol. If true, anyone can pull a DocumentArray if they know its name.
+            Setting this to false will restrict access to only the creator.
+        :param show_progress: If true, a progress bar will be displayed.
+        :param branding: Only used by ``jinaai`` protocol. A dictionary of branding information to be sent to Jina AI Cloud. {"icon": "emoji", "background": "#fff"}
+        """
+        logging.info(f'Pushing {len(self)} docs to {url}')
+        protocol, name = url.split('://', 2)
```

Comment:
Can we put this `url.split('://', 2)` into a static function ? I saw that we are using it a couple of times

Reply:
Hrmm ok. I figured it was quite short and not too hard to understand. But the literal typing requires this now so its done.